modules/core/04-channel/keeper/upgrade.go

```diff
@@ -109,6 +110,110 @@ func (k Keeper) WriteUpgradeTryChannel(
+	portID,
+	channelID string,
+	proposedUpgradeFields types.UpgradeFields,
+	counterpartyChannel types.Channel,
+	counterpartyUpgrade types.Upgrade,
+	proofCounterpartyChannel,
+	proofCounterpartyUpgrade []byte,
+	proofHeight clienttypes.Height,
+) error {
+	channel, found := k.GetChannel(ctx, portID, channelID)
+	if !found {
+		return errorsmod.Wrapf(types.ErrChannelNotFound, "port ID (%s) channel ID (%s)", portID, channelID)
+	}
+
+	connection, err := k.GetConnection(ctx, channel.ConnectionHops[0])
+	if err != nil {
+		return errorsmod.Wrap(err, "failed to retrieve connection using the channel connection hops")
+	}
+
+	if connection.GetState() != int32(connectiontypes.OPEN) {
+		return errorsmod.Wrapf(connectiontypes.ErrInvalidConnectionState, "connection state is not OPEN (got %s)", connectiontypes.State(connection.GetState()).String())
+	}
+
+	// verify the counterparty channel state containing the upgrade sequence
+	if err := k.connectionKeeper.VerifyChannelState(
+		ctx,
+		connection,
+		proofHeight, proofCounterpartyChannel,
+		channel.Counterparty.PortId,
+		channel.Counterparty.ChannelId,
+		counterpartyChannel,
+	); err != nil {
+		return errorsmod.Wrap(err, "failed to verify counterparty channel state")
+	}
+
+	// verifies the proof that a particular proposed upgrade has been stored in the upgrade path of the counterparty
+	if err := k.connectionKeeper.VerifyChannelUpgrade(
+		ctx,
+		connection,
+		proofHeight, proofCounterpartyUpgrade,
+		channel.Counterparty.PortId,
+		channel.Counterparty.ChannelId,
+		counterpartyUpgrade,
+	); err != nil {
+		return errorsmod.Wrap(err, "failed to verify counterparty upgrade")
+	}
+
+	// the current upgrade handshake must only continue if both channels are using the same upgrade sequence,
+	// otherwise an error receipt must be written so that the upgrade handshake may be attempted again with synchronized sequences
+	if counterpartyChannel.UpgradeSequence != channel.UpgradeSequence {
+		// save the previous upgrade sequence for the error message
+		prevUpgradeSequence := channel.UpgradeSequence
+
+		// error on the higher sequence so that both chains synchronize on a fresh sequence
+		channel.UpgradeSequence = sdkmath.Max(counterpartyChannel.UpgradeSequence, channel.UpgradeSequence)
+		k.SetChannel(ctx, portID, channelID, channel)
+
+		return types.NewUpgradeError(channel.UpgradeSequence, errorsmod.Wrapf(
+			types.ErrIncompatibleCounterpartyUpgrade, "expected upgrade sequence (%d) to match counterparty upgrade sequence (%d)", prevUpgradeSequence, counterpartyChannel.UpgradeSequence),
+		)
+	}
+
+	// assert that both sides propose the same channel ordering
+	if proposedUpgradeFields.Ordering != counterpartyUpgrade.Fields.Ordering {
+		return types.NewUpgradeError(channel.UpgradeSequence, errorsmod.Wrapf(
+			types.ErrIncompatibleCounterpartyUpgrade, "expected upgrade ordering (%s) to match counterparty upgrade ordering (%d)", proposedUpgradeFields.Ordering, counterpartyUpgrade.Fields.Ordering),
```

Comment:
I think if %d is used here the enums int32 value will be formatted into the string, right?

```suggestion
			types.ErrIncompatibleCounterpartyUpgrade, "expected upgrade ordering (%s) to match counterparty upgrade ordering (%s)", proposedUpgradeFields.Ordering, counterpartyUpgrade.Fields.Ordering),
```